### DISCUSSION TITLE
How can we measure progress?

### DISCUSSION THREAD

**Discussion by clizbe** [TulipaEnergy/TulipaEnergyModel.jl#307] [Questions]
Our internal leadership wants to keep better tabs on our progress.
To me this is tricky considering our organic/iterative development - but I agree it would be nice to recognize early if we need to triage.

One way might be defining Key Performance Indicators (KPIs). These can optionally have a Threshold (requirement) and a Target (bonus).
Do you have ideas about what we could use for KPIs?
I thought maybe:
- Speed of building model of European Case
- Speed of solving European Case
- Number of features implemented?
- Size of problem solvable?
- Something to do with data handling?
- Something to do with pre-processing?
- Something to do with documentation?

To me KPIs are a nice way of focusing on quality over quantity. We can also make a list of core features and track as they're delivered, but I think including quality somewhere is important as well.

I thought maybe eSciencers would have ideas about this since you have more experience with programming projects.
@suvayu @abelsiqueira @lsoucasse @datejada @gnawin

**Comment by suvayu** [comment]
KPIs are a good idea, and those are *all* valid candidates.  I think maybe we can treat them as categories, and within each category define KPIs under required & target, where we can choose between them based on how often that step needs to be done.  So something along the lines of this maybe?

| category | required | target |
| --- | --- | --- |
| performance | benchmark model: Europe | |
|                      | - model building < 1m         | |
|                      | - model solve < 1h           | |
| model features |  | |
| problem size | (how to define this?) | |
| data pipeline | - no data duplication | - explorable datasets |
|                      | - handle scenarios | |
|                      | - handle solver options | |
|                      | - customisable pre-processing | |
| documentation | | |

**Reply by clizbe** [reply]
For problem size, I was just thinking number of variables & constraints. 
As in, how large of a problem can we handle? But I guess that might depend too much on the type of problem.

**Reply by suvayu** [reply]
If we can relate the number of variables & constraints to something more "visible" I think it could work well. While on it's own it is a valid criteria, unless you know the model formulation inside out, it's hard to judge if a particular approach leads to optimum variables & constraints or not.  I guess this should be part of the model formulation documentation (#292).  Also, we should emit this as diagnostic information before starting a model solve.  We could also add a warning if this is too large.

**Reply by clizbe** [reply]
| category | name | required | target |
| --- | --- | --- | --- |
| performance | | benchmark model: European Case | |
|                      | Model build time | < 1m         | |
|                      | Model solve time | < 1h (with tradeoffs?) | < 1h (without tradeoffs?) |
| model features | *Expand list from PPT | Which are required | Which are desired |
| capability | Solvable problem size | # nodes, # carriers, time, features? (maybe hard to measure) | |
| data pipeline | Datasets | No data duplication - *what does this mean?* @suvayu   | Explorable datasets |
|                      | Scenario specification | Scenario parameter specification without changing source data | UI for scenarios |
|                      | Solver options | User can specify | UI for solver options |
|                      | Pre-processing | All parts integrated | Customisable pipeline |
| pre-processing | *List of methods? | Which are required | Which are desired |
| documentation | Code | All functions explained | 1: Code diagram 2: Connection with formulation |
| | Formulation | Tulipa formulation | Citations to all related publications |
| | How to Use | How to get started (download & run) | 1: Demo case study 2: Tutorials/How-tos for specific features
| graphs | Graph builder | Core graphs generate automatically | Customizable graphs (UI) |
| | Results sharing | None | Explorable graphs online |

**Reply by clizbe** [reply]
That's a really good point that if we're doing it right, the size of the problem should be smaller than what other models are solving.
Maybe instead, the size of the grid and detail somehow... Like # energy carriers, length of solve (year), etc?

**Reply by clizbe** [reply]
As in, the size of the scenario, rather than how large it is when we go to solve it.
But maybe that's a hollow pursuit and we should just focus on the benchmark case.

**Reply by suvayu** [reply]
> Maybe instead, the size of the grid and detail somehow... Like # energy carriers, length of solve (year), etc?

Yes, something like this.  But there are some hidden caveats, for example time resolution of the profiles,  I think for performance comparisons, we can pick the simplest case for these kinds of factors, and state them explicitly.  That should cover it

> As in, the size of the scenario, rather than how large it is when we go to solve it.

Not sure I follow.  Do you mean to say include performance numbers for a multi-scenario run of the benchmark model?  In that case I would say no.  Since it doesn't add any uniquely new information.

**Reply by suvayu** [reply]
> No data duplication - what does this mean?

Avoid the current "copy-paste" datasets to run a different scenario issue.  This also could extend to how the datasets are handled for multiple scenarios inside Julia, e.g. shared arrays when the dataset doesn't change.  And at the most extreme end of the wishlist is something like "Gather Update Solve Scatter".

**Reply by clizbe** [reply]
I think this is a good start! I'm going to swing it by management and if they're cool with it we'll solidify it. They might still want some kind of timeline, but I'm hoping just being able to SEE our progress at any point in time will be enough.

**Comment by clizbe** [comment]
Now we have a great planning! See the details of the Project.
